### PR TITLE
[SPARK-44424][CONNECT][PYTHON] Python client for reattaching to existing execute in Spark Connect

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to implement the Python client side for https://github.com/apache/spark/pull/42228.

Basically this PR applies the same changes of `ExecutePlanResponseReattachableIterator`, and `SparkConnectClient` to PySpark as  the symmetry.

### Why are the changes needed?

To enable the same feature in https://github.com/apache/spark/pull/42228

### Does this PR introduce _any_ user-facing change?

Yes, see https://github.com/apache/spark/pull/42228.

### How was this patch tested?

Existing unittests because it enables the feature by default. Also, manual E2E tests.

### CODE REVIEW

File: python/pyspark/sql/connect/client/reattach.py
Comment (juliuszsompolski) [inline python/pyspark/sql/connect/client/reattach.py:121]: is it possible to make the ReleaseExecute async - the idea is that client just "fires-and-forget" them as help to server to not do too much buffering, but doesn't get blocked on them.

File: python/pyspark/sql/connect/client/reattach.py
Comment (juliuszsompolski) [inline python/pyspark/sql/connect/client/reattach.py:155]: heads up: changed to not have release_type enum + a oneof, but just a single oneof.

File: python/pyspark/sql/connect/client/reattach.py
Comment (juliuszsompolski) [inline python/pyspark/sql/connect/client/reattach.py:102]: heads up: ResponseComplete renamed to ResultComplete

**Comment by juliuszsompolski** [comment]
It's tracked by https://issues.apache.org/jira/browse/SPARK-44424 already as well.

**Review by juliuszsompolski — APPROVED** [review]
Looks good to me, but someone knowing more python should review.

Question: There are python test that work E2E, i.e. connect the client with a separate server?
Could we pass configs like in https://github.com/apache/spark/pull/42228/files#diff-c389e865d11bc54f29d61207c50ced1f9561de1488865eaf11caf789dd6003c2R144 for the testing, so that E2E testing exercises reattaching?

File: python/pyspark/sql/connect/client/reattach.py
Comment (juliuszsompolski) [inline python/pyspark/sql/connect/client/reattach.py:223]: on close, it should do release_all if it wasn't done.
in scala the Iterator unfortunately doesn't have a close, but if python has it, it should do it.

File: python/pyspark/sql/connect/client/reattach.py
Comment (grundprinzip) [inline python/pyspark/sql/connect/client/reattach.py:116]: attempt is of type `AttemptManager` which contains a `_retry_state` that has a _count. We can make the retry state visible in the attempt manager if you want to avoid having a first try variable

File: python/pyspark/sql/connect/client/reattach.py
Comment (grundprinzip) [inline python/pyspark/sql/connect/client/reattach.py:120]: or maybe `attempt.first_try` (added to attempt manager)

File: python/pyspark/sql/connect/client/reattach.py
Comment (grundprinzip) [inline python/pyspark/sql/connect/client/reattach.py:83]: So in the case of the target resource throwing a retriable error, this will already raise an exception and break the retry logic. It looks like the current retry handling can only deal with subsequent errors in the handling of the response stream but not the initial error itself.

Comment (grundprinzip) [reply python/pyspark/sql/connect/client/reattach.py:83]: @juliuszsompolski is this the same for the current scala client?

File: python/pyspark/sql/connect/client/reattach.py
Comment (grundprinzip) [inline python/pyspark/sql/connect/client/reattach.py:142]: Just readability.

```suggestion
                    if not self._result_complete and not has_next:
```

File: python/pyspark/sql/connect/client/reattach.py
Comment (grundprinzip) [inline python/pyspark/sql/connect/client/reattach.py:144]: This could raise itself with a retriable error.

File: python/pyspark/sql/connect/client/reattach.py
Comment (juliuszsompolski) [reply python/pyspark/sql/connect/client/reattach.py:83]: At least in scala, calling the `executePlan` never throws error, even if the service is absolutely not available and one could expect immediate error; I tested that with completely killing the server, then calling executePlan, and never getting an actual error before trying to call next()... The error is only thrown on the first `next()` or `hasNext()` call of the resulting iterator. So in scala not catching and retrying here, but at the first iterator invocation is fine. So in https://github.com/apache/spark/pull/42228/files#diff-2efc03ab3edd5e5f78652b1454b31fd4917113de09a64fbdc7352f150c685628R89 and also in https://github.com/apache/spark/pull/41829/files#diff-79d952ace19b8b9d4b04e010653aba35e4aeaca957946cc68414b7a766deea99R68 doesn't wrap the first call
we do not retry.

But it's a good callout that it needs checking if python GRPC behaves the same, or maybe just defensively wrap it in retry anyway (and do the same in scala just in case it's a GRPC internal quirk that we shouldn't depend on), or at least it deserves a comment.

Comment (juliuszsompolski) [reply python/pyspark/sql/connect/client/reattach.py:83]: I raised https://github.com/apache/spark/pull/42274 to be defensive about it in scala.

File: python/pyspark/sql/connect/client/core.py
Comment (juliuszsompolski) [reply python/pyspark/sql/connect/client/core.py:646]: This is used only when creating ExecutePlanResponseReattachableIterator, and once that's created, it will know what to do even if this gets flipped.

File: python/pyspark/sql/connect/client/reattach.py
Comment (juliuszsompolski) [reply python/pyspark/sql/connect/client/reattach.py:83]: ... but what I write here implies that if the error was indeed during the initial ExecutePlan that didn't reach the server, and we catch it on the first hasNext() later, then Reattach won't work, because the query won't be on the server, so it will fail with INVALID_HANDLE.OPERATION_NOT_FOUND. So this indeed means that the case it won't work if retry of the initial ExecutePlan is needed, and wrapping the retry here won't help for that.

Comment (juliuszsompolski) [reply python/pyspark/sql/connect/client/reattach.py:83]: retrying the failure of initial operation starting is tricky, because:
- if the initial operation reached the server, retry of ExecutePlan would fail with INVALID_HANDLE.OPERATION_ALREADY_EXISTS; in that case ReattachPlan should be used. And this is what happens here.
- but if the initial operation didn't reach the server, then the ReattachPlan will fail with INVALID_HANDLE.OPERATION_NOT_FOUND. This could be improved by making it switch to retrying the ExecutePlan then
But I think this is a border case that can be a followup or decided it's not worth it.

I believe in the state as is, we will get an INVALID_HANDLE.OPERATION_NOT_FOUND error in case the original ExecutePlan never reached the server. In that case, we could improve it by trying to call ExecutePlan again. But that would have to be caught not here, but during the next() calls.

Comment (juliuszsompolski) [reply python/pyspark/sql/connect/client/reattach.py:83]: So now I have 3 PRs:
* https://github.com/apache/spark/pull/42281 adds some comments in scala to explain that GRPC throws errors deferred to the iterator functions, and the limitation of retry of the original execute didn't reach the server
* https://github.com/apache/spark/pull/42282 fixes the retry in case original execute didn't reach the server
* https://github.com/apache/spark/pull/42274 does that plus does overkill retries in case the assumption that calling executePlan and reattachExecute won't throw is wrong.

I would suggest merging the first, and discussing this corner case later.

File: python/pyspark/sql/connect/client/reattach.py
Comment (HyukjinKwon) [reply python/pyspark/sql/connect/client/reattach.py:108]: it causes a circular import (because `SparkConnectClient` also imports `ExecutePlanResponseReattachableIterator`)

File: python/pyspark/sql/connect/client/reattach.py
Comment (HyukjinKwon) [reply python/pyspark/sql/connect/client/reattach.py:99]: Actually I just used a different method to make this clear.

File: python/pyspark/sql/connect/client/reattach.py
Comment (HyukjinKwon) [reply python/pyspark/sql/connect/client/reattach.py:116]: Yup, this is the time that this context manager shines :-)

File: python/pyspark/sql/connect/client/reattach.py
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/reattach.py:185]: I can't find the symmetry way of doing StreamObserver in Python. I know this isn't super nice but it would at least work.

Comment (juliuszsompolski) [reply python/pyspark/sql/connect/client/reattach.py:185]: This is fine. Maybe would have been even cleaner this way in scala, because the current code there of invoking an async execution, and then chaining another async execution from the onError of it is actually a bit funky.
A thread with retry loop over sync requests would be simpler code.

File: python/pyspark/testing/connectutils.py
Comment (juliuszsompolski) [inline python/pyspark/testing/connectutils.py:173]:
```suggestion
        # Make the server terminate reattachable streams every 1 second and 123 bytes,
        # to make the tests exercise reattach.
        conf.set("spark.connect.execute.reattachable.senderMaxStreamDuration", "1s")
        conf.set("spark.connect.execute.reattachable.senderMaxStreamSize", "123")
```

**Comment by HyukjinKwon** [comment]
(oops mistakenly removed the comment after applying the change)

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.5.

File: python/pyspark/testing/connectutils.py
Comment (allisonwang-db) [inline python/pyspark/testing/connectutils.py:176]: @HyukjinKwon I saw this error when running unit tests locally: `org.apache.spark.sql.AnalysisException: Cannot modify the value of a static config: spark.connect.execute.reattachable.senderMaxStreamDuration.`

Comment (HyukjinKwon) [reply python/pyspark/testing/connectutils.py:176]: That should be fixed in the lastest master branch.

**Pull request closed without merging**
